src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,185 @@
+                  });
+            });
+
+            inner_hash.modify(
+              it_hash, [next](timestamped_user& item) { item.client = next; });
+
+            // Add a prefix to the evicted key so debugging logs
+            // is easier
+            _evicted_items.emplace_back(
+              k + random_generators::gen_alphanum_string(8), old);
+        } else {
+            vlog(plog.debug, "Reuse client for user {}", k);
+        }
+
+        client = it_hash->client;
+
+        // Convert hash iterator to list iterator
+        auto it_list = _cache.project<underlying_list>(it_hash);
+
+        // Update the timestamp and relocate the element
+        // to the front
+        inner_list.modify(it_list, [](timestamped_user& item) {
+            item.last_used = timestamped_user::clock::now();
+        });
+        inner_list.relocate(inner_list.begin(), it_list);
+    }
+
+    co_return client;
+}
+
+namespace {
+template<typename List_Type>
+ss::future<> clean_clients_from_list(
+  List_Type& list, mutex& lock, std::chrono::milliseconds keep_alive) {
+    auto has_expired = timestamped_user::clock::now();
+    auto condition = [keep_alive, has_expired](const timestamped_user& item) {
+        return has_expired >= (item.last_used + keep_alive);
+    };
+
+    auto first = list.begin();
+    auto last = list.end();
+    while (first != last) {
+        if (condition(*first)) {
+            vlog(plog.debug, "Remove {} from cache", first->key);
+            // Stop client before erase
+            co_await lock.with([&first] {
+                return first->client->stop().handle_exception(
+                  [](std::exception_ptr ex) {
+                      // The stop failed
+                      vlog(
+                        plog.debug,
+                        "Stale client stop already happened {}",
+                        ex);
+                  });
+            });
+            list.erase(first++);
+        } else {
+            ++first;
+        }
+    }
+}
+
+template<typename List_Type>
+ss::future<> stop_clients_until_empty(List_Type& list, mutex& lock) {
+    while (!list.empty()) {
+        co_await lock.with([&list] {
+            return list.back().client->stop().handle_exception(
+              [](std::exception_ptr ex) {
+                  // The stop failed
+                  vlog(plog.debug, "Stop already happened {}", ex);
+              });
+        });
+        list.pop_back();
+    }
+}
+} // namespace
+
+// Modified boost remove_if so we can call kafka::client::stop
+// See sequenced_index_remove in
+// https://github.com/boostorg/multi_index/blob/develop/include/boost/multi_index/detail/seq_index_ops.hpp
+void kafka_client_cache::clean_stale_clients() {
+    auto& inner_list = _cache.get<underlying_list>();
+    ssx::background = clean_clients_from_list(inner_list, _lock, _keep_alive);
+    ssx::background = clean_clients_from_list(
+      _evicted_items, _lock, _keep_alive);
+}
```

Comment:
Doing this work in the background negates the dance we did earlier with `timer.arm`. I think this should return a future and await the cleaning.

Reply:
Removed background and used co_await instead.